mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala

```diff
@@ -32,20 +32,33 @@ import org.apache.spark.util.random.XORShiftRandom
  *
  */
 private[ann] trait Layer extends Serializable {
+
+  /**
+   * Number of weights that is used to allocate memory for the weights vector
+   */
+  val weightSize: Int
+
+  /**
+   * Returns the output size given the input size (not counting the stack size).
+   * Output size is used to allocate memory for the output.
+   * @param inputSize input size
+   * @return output size
+   */
+  def outputSize(inputSize: Int): Int
+
   /**
    * Returns the instance of the layer based on weights provided
    * @param weights vector with layer weights
-   * @param position position of weights in the vector
    * @return the layer model
    */
-  def getInstance(weights: Vector, position: Int): LayerModel
-
+  def instance(weights: BDV[Double]): LayerModel
```

Comment:
`Layer` is really implementing the Factory pattern for `LayerModel`. Should we consider renaming it to use the proper name?